File: frontends/kde/KeyForward.cpp
/*
 * KeyForward -- see KeyForward.h.
 *
 * Copyright (C) 2026 Thomas Cherryhomes
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include "KeyForward.h"

#include "a2600session.h"

uint32_t a2600KeysymFromQt(const QKeyEvent *e)
{
    /* On X11 and Wayland Qt reports the X keycode, which is the evdev code
     * plus 8. */
    const quint32 sc = e->nativeScanCode();
    if (sc >= 8) {
        const uint32_t ks = a2600session_keysym_from_evdev(sc - 8);
        if (ks) return ks;
    }

    switch (e->key()) {
    case Qt::Key_Up:      return A2600_KEYSYM_UP;
    case Qt::Key_Down:    return A2600_KEYSYM_DOWN;
    case Qt::Key_Left:    return A2600_KEYSYM_LEFT;
    case Qt::Key_Right:   return A2600_KEYSYM_RIGHT;
    case Qt::Key_Tab:     return A2600_KEYSYM_TAB;
    case Qt::Key_Escape:  return A2600_KEYSYM_ESCAPE;
    case Qt::Key_Return:  return A2600_KEYSYM_RETURN;
    case Qt::Key_Enter:   return A2600_KEYSYM_KP_ENTER;
    case Qt::Key_Backspace: return A2600_KEYSYM_BACKSPACE;
    case Qt::Key_Space:   return A2600_KEYSYM_SPACE;
    case Qt::Key_Shift:   return A2600_KEYSYM_LSHIFT;
    case Qt::Key_Control: return A2600_KEYSYM_LCTRL;
    case Qt::Key_Alt:     return A2600_KEYSYM_LALT;
    case Qt::Key_Asterisk: return '*';
    case Qt::Key_NumberSign: return '#';
    default: break;
    }
    const int k = e->key();
    if (k >= Qt::Key_F1 && k <= Qt::Key_F12)
        return A2600_KEYSYM_F1 + (k - Qt::Key_F1);
    if (k >= Qt::Key_0 && k <= Qt::Key_9)
        return (uint32_t)('0' + (k - Qt::Key_0));
    if (k >= Qt::Key_A && k <= Qt::Key_Z)
        return (uint32_t)('a' + (k - Qt::Key_A));
    if (k >= 0x20 && k <= 0x7e)
        return (uint32_t)k;
    return 0;
}

File: frontends/macos/KeyForward.m
/*
 * KeyForward -- see the header.
 *
 * Copyright (C) 2026 Thomas Cherryhomes
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "KeyForward.h"

#import <Carbon/Carbon.h>   /* kVK_* virtual key codes */

#include "a2600session.h"

uint32_t A2600KeysymFromEvent(NSEvent *event)
{
    const uint32_t ks = a2600session_keysym_from_macos_keycode([event keyCode]);
    if (ks) return ks;

    switch ([event keyCode]) {
    case kVK_UpArrow:    return A2600_KEYSYM_UP;
    case kVK_DownArrow:  return A2600_KEYSYM_DOWN;
    case kVK_LeftArrow:  return A2600_KEYSYM_LEFT;
    case kVK_RightArrow: return A2600_KEYSYM_RIGHT;
    case kVK_Tab:        return A2600_KEYSYM_TAB;
    case kVK_Escape:     return A2600_KEYSYM_ESCAPE;
    case kVK_Return:     return A2600_KEYSYM_RETURN;
    case kVK_Space:      return A2600_KEYSYM_SPACE;
    case kVK_Delete:     return A2600_KEYSYM_BACKSPACE;
    default: break;
    }

    /* Anything the table does not know: the character typed. */
    NSString *chars = [event charactersIgnoringModifiers];
    if ([chars length] == 1) {
        const unichar c = [chars characterAtIndex:0];
        if (c >= 0x20 && c <= 0x7e) {
            if (c >= 'A' && c <= 'Z') return (uint32_t)(c - 'A' + 'a');
            return (uint32_t)c;
        }
        if (c >= NSF1FunctionKey && c <= NSF12FunctionKey)
            return A2600_KEYSYM_F1 + (uint32_t)(c - NSF1FunctionKey);
    }
    return 0;
}

/* Modifier keys arrive as flagsChanged, not keyDown/keyUp. The key code
 * says which key; the corresponding flag says whether it went down. */
uint32_t A2600KeysymFromFlagsChange(NSEvent *event, int *down)
{
    const NSEventModifierFlags f = [event modifierFlags];
    switch ([event keyCode]) {
    case kVK_Shift:        *down = (f & NSEventModifierFlagShift) != 0;   return A2600_KEYSYM_LSHIFT;
    case kVK_RightShift:   *down = (f & NSEventModifierFlagShift) != 0;   return A2600_KEYSYM_RSHIFT;
    case kVK_Control:      *down = (f & NSEventModifierFlagControl) != 0; return A2600_KEYSYM_LCTRL;
    case kVK_RightControl: *down = (f & NSEventModifierFlagControl) != 0; return A2600_KEYSYM_RCTRL;
    /* Option is where a Mac keyboard puts the key a PC calls Alt. */
    case kVK_Option:       *down = (f & NSEventModifierFlagOption) != 0;  return A2600_KEYSYM_LALT;
    case kVK_RightOption:  *down = (f & NSEventModifierFlagOption) != 0;  return A2600_KEYSYM_RALT;
    default: *down = 0; return 0;
    }
}

NSColor *A2600AccentColor(void)
{
    return [NSColor colorWithSRGBRed:((A2600SESSION_ACCENT_RGB >> 16) & 0xff) / 255.0
                               green:((A2600SESSION_ACCENT_RGB >> 8) & 0xff) / 255.0
                                blue:(A2600SESSION_ACCENT_RGB & 0xff) / 255.0
                               alpha:1.0];
}
